[Site Editor - Page Inspector]: Add ability to switch templates

ntsekouras opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Resolves: https://github.com/WordPress/gutenberg/issues/51347

This is the first iteration to add the ability to switch templates in site editor's page inspector. Currently we show the `swap template` button if we have custom templates. Additionally we skip showing the button if we have only one custom template that is already assigned to the page.


## How?

For showing the templates list we use the `BlockPatternsList` component that handles many things already, like keyboard navigation, and for styling consistency with similar modal. Also in the future we could suggest patterns and create custom templates on the fly.

The challenging thing for me was that while we need to edit one entity(`page`), we also need to update the current block editors' edited entity(`template`). That's why I added the `setEditedPost` action, to update both at once. Furthermore the `undo` functionality was added through the `snackbar` because the specific `undo` needs a side effect(to update the edited entity). I'm open to suggestions for alternatives or something better. 





## Testing Instructions
1. Observe that with no custom templates or a single template assigned to the specific page being tested, the swap button doesn't appear
2. Have multiple custom templates and by clicking the `swap template` button the changes are applied and the `undo` in snackbar works properly.



## Screenshots or screencast <!-- if applicable -->


https://github.com/WordPress/gutenberg/assets/16275880/de612ac5-397b-43cc-9097-ae965e9a14c8

youknowriad commented:
Do you think we should be doing the same in the post editor, in the "template" panel as well. These seem like they should do the exact same thing?

ntsekouras commented:
>Do you think we should be doing the same in the post editor, in the "template" panel as well

You mean the templates list, right? In post editor we have the same functionality with a select control in `page` panel though. The two panels(when we `edit template` in post editor and in site editor) have diverged and probably makes more sense to focus on template building/flows in site editor, no?

youknowriad commented:
Yeah, my question essentially is why they have diverged, and should the swapping be similar in the post editor (with a visual list instead of a select)

youknowriad commented on docs/reference-guides/data/data-core-edit-site.md:
```diff
@@ -268,6 +268,18 @@ _Parameters_
 -   _options_ `[Object]`:
 -   _options.allowUndo_ `[boolean]`: Whether to allow the user to undo reverting the template. Default true.
 
+### setEditedPost
```
Were the existing actions `setPage` / `setTemplate` not enough for the current use case here?

ntsekouras replied:
`setTemplate` sets as edited entity the template and loads that specific view(in comparison to the page view with the resolved template for that page).

I tried this with both of the above and with `setPage` we don't have something to trigger the new template assigned and set it as the edited entity(as the postId and postType remain the same). I might have missed something though 🤷 .

youknowriad replied:
Ok. I actually think the problem comes from how we're structuring our reducer. 

In other words, setting the template of a post, should be enough to tell the site editor to actually reload with the new template. The problem is that we have this "state" in two places so we need to synchronize.

In an ideal scenario, I think the site editor reducer (state) shouldn't contain any reference to the "template". It should only contain a reference the main edited entity (the one in the url). So for instance when a page or post is opened, it should only contain a reference to the postId and postType. And the template should be retrieved using regular "core-data" selectors when needed.

That said, this is definitely out of scope of the current PR and something I've noticed before. I think we would benefit (simplification) from such a refactoring at some point.

---

That said, it's a bit weird that `setPage` doesn't take the edited entity into consideration.

Mamaduka replied:
I had the same question. Besides the extra requests that `setPage` does, the actions look the same.

P.S. Should we stick with the current naming convention and use `setEditedPage` as the name?

ntsekouras replied:
>That said, it's a bit weird that setPage doesn't take the edited entity into consideration.

After trying things until I get the flow it seems the issue was with `resolveSelect` inside `setPage` action.
For now I replaced it with `apiFetch` call. Any better suggestions? 🤔

youknowriad commented on packages/edit-site/src/components/sidebar-edit-mode/page-panels/swap-template-button.js:
```diff
@@ -0,0 +1,151 @@
+	const shownTemplates = useAsyncList( templatesAsPatterns );
+	const onClickPattern = async ( template ) => {
+		entitiy.edit( { template: template.name }, { undoIgnore: true } );
+		await entitiy.save();
```
Not necessarily specific to this PR: I noticed that this pattern `edit` with undoIgnore and `save` is kind of repeated in a lot of places in our code base. I wonder if we should just use the "save" directly for that (I believe the save can receive edits too) or some editAndSave action.

youknowriad commented on packages/edit-site/src/components/sidebar-edit-mode/page-panels/style.scss:
```diff
@@ -5,6 +5,37 @@
+}
+
+
+.edit-site-page-panels__swap-template__modal-content .block-editor-block-patterns-list {
```
If I'm not wrong, this CSS is repeated in a lot of places, I wonder if we should have this in the component itself (like gridColumns prop or something like that)

ntsekouras replied:
That's true, but I focused for now with the technical implementation for the first round of reviews.

Mamaduka commented:
> Do you think we should be doing the same in the post editor, in the "template" panel as well.

The post editor selector also works for classic themes, where a similar template list doesn't make sense, as we can't display previews for PHP templates.

Mamaduka commented on packages/edit-site/src/components/sidebar-edit-mode/page-panels/swap-template-button.js:
```diff
@@ -0,0 +1,151 @@
+			getEditedPostType(),
+			getEditedPostId()
+		);
+	}, [] );
```
Suggestion (non-blocking): The component only needs the data to check if there are more templates for selection. Maybe we should derive this data inside the `mapSelect` and avoid passing `templates` via props.

Mostly in the spirit of Redux and selecting minimal data required for a component :)

ntsekouras replied:
What do you mean with `inside mapSelect`?

Mamaduka commented on packages/edit-site/src/components/sidebar-edit-mode/page-panels/swap-template-button.js:
```diff
@@ -0,0 +1,151 @@
+
+function TemplatesList( { templates, currentTemplate, onSelect } ) {
+	const { postId, postType } = useSelect( ( select ) => {
+		const _context = select( editSiteStore ).getEditedPostContext();
```
Question (unrelated): The selector looks to be deprecated (#46433) but is still used in a few places in the site editor. Should we un-deprecate it before there's a better alternative?

ntsekouras replied:
Hm.. @youknowriad it seems even in your PR that deprecated, you still used it and there doesn't seem to be an alternative for that.. Any insights?

youknowriad replied:
I think I probably deprecated it and changed my mind (too big of a refactor) and forgot about the deprecation.

jameskoster commented:
Thanks @ntsekouras. For some reason I'm only seeing one template as an option, despite several custom templates existing. And when I swap to that template, I can no longer swap back 🤔 

After swapping, the template is no longer listed in the details panel either. Quick video to demonstrate these issues:

https://github.com/WordPress/gutenberg/assets/846565/dfa41bc8-dbbf-4d18-ba81-83e492543593

A couple of nice-to-haves that can be addressed in follow-ups if need be:

* Could we include the currently active template in the modal, and visually highlight it (blue outline)?
* Could we also display the template names in the modal? Sometimes templates can be quite similar in appearance, and the name can help distinguish between them.
* Instead of "Template swapped" in the Snackbar, perhaps a more explicit "[Template name] applied" would be helpful.

Also leaving a note that the template panel itself could use some holistic design attention, and that we should circle back to that as part of the polish effort in the lead up to 6.3.

jameskoster commented:
One other observation: swapping the template seems to get saved instantly? I don't think that should be the case.

ntsekouras commented:
>One other observation: swapping the template seems to get saved instantly? I don't think that should be the case.

I'm not sure how easy that would be because of:
>Furthermore the undo functionality was added through the snackbar because the specific undo needs a side effect(to update the edited entity).

That's the same problem the other way around.

>I'm only seeing one template as an option, despite several custom templates existing

Hm.. that's my logic's bug. I'll see to fix it. Write now I was fetching only the custom templates created by users.

jameskoster commented:
Would removing the Undo affordance help? I don't think we need that if the change isn't immediately published.

If there's no way around it then we'll need to adjust the modal design a bit to make it obvious that the change will be published straight away. It would likely involve adding a footer with "Save" button. Let me know and I'll create a mockup.

Seems like you're working on this at the moment but just wanted to note that:

* I now see theme-supplied custom templates in the modal 🎉
* I still do not see the default template (e.g. Page) as an option when a custom template is selected.
* The 'template' row is still missing on the Page details panel when a custom template is applied.

ntsekouras commented:
>If there's no way around it then we'll need to adjust the modal design a bit to make it obvious that the change will be published straight away

I cannot see a way around it right now, so we could explore some designs for that, yes.

>I still do not see the default template (e.g. Page) as an option when a custom template is selected.

I'll see what I can do about that. This involves the template hierarchy and I don't think we have something ready right now. For example if we have both `page` and `singular` templates we should probably show only `page`, right? Also do we need to take into account `posts` or other custom post types?

>The 'template' row is still missing on the Page details panel when a custom template is applied.

I'll check if this is related to this PR or on trunk too.

youknowriad commented:
> I'll see what I can do about that. This involves the template hierarchy and I don't think we have something ready right now. For example if we have both page and singular templates we should probably show only page, right? Also do we need to take into account posts or other custom post types?

I think in the post editor we just name it "default" or something like that to work around this issue. In other words, do you think this PR could use the "same code" (even if copy/pasted) to get the list of templates as the code we already have in the post editor?

jameskoster commented:
> For example if we have both page and singular templates we should probably show only page, right?

That's correct. No point showing singular in that case as it wouldn't be possible to actually apply that template. But if page didn't exist, then singular would be there. A generic "Default" label like Riad suggested could work in the short term.

> Also do we need to take into account posts or other custom post types?

How do you mean? Display templates from other post types as options in this modal? I don't think we need to do that initially – it feels like something that would be part of a new template flow. E.g. you might clone single post when creating page-about.

> I cannot see a way around it right now, so we could explore some designs for that, yes.

I'll work on that and report back.

jameskoster commented:
> I'll work on that and report back.

Here are a couple of options, I don't have a strong feeling about either, so the decision will likely come down to which is easier to implement / maintain.

**Option 1**

<img width="2432" alt="Choose template with save" src="https://github.com/WordPress/gutenberg/assets/846565/0d3c2c77-af78-4ad5-8d02-b6eece4b42be">

* A notice informs the user that swapping a template is a change that will be published immediately
* The current template is visible and selected when you open the modal
* ‘Swap’ button disabled until a different template is selected

**Option 2**

<img width="1851" alt="Two step" src="https://github.com/WordPress/gutenberg/assets/846565/1c0046f0-fff5-4213-8a88-857280629319">

The initial modal remains the same as it is now, but when you select a template a subsequent modal appears asking if you want to update the page immediately.

ntsekouras commented:
>I think in the post editor we just name it "default" or something like that to work around this issue. In other words, do you think this PR could use the "same code" (even if copy/pasted) to get the list of templates as the code we already have in the post editor?

I've been trying this yesterday for a while and the use case is different. This is because in order to apply the default template we just need to update the page with a value of `template: ''`. This is not good enough for us, because we need the template to have resolved and use more info. The same problem exists in post editor too, though not so visible, when we update from a custom template to the `default` - then the resolution doesn't happen without a reload, but doesn't matter so much there. 

I'll try to see if I can solve this with my initial explorations which were to just need to update the `page` and the site editor should reload with the new edited entity.

---- 

--edit

I pushed some changes to resolve the default template after selecting the `default`, but still no way to have the preview(blocks) of that. I'll continue to explore what I'm mentioning above, but not sure it will make it for 6.3 at this point..

youknowriad commented on packages/edit-site/src/store/actions.js:
```diff
@@ -260,9 +246,11 @@ export const setPage =
-		const template = await registry
-			.resolveSelect( coreStore )
-			.__experimentalGetTemplateForLink( page.path );
+		const { data: template } = await apiFetch( {
```
Did you do that because of the caching?

youknowriad replied:
What about tweaking `__experimentalGetTemplateForLink.shouldInvalidate` to invalidate the cache at the right moment?

ntsekouras replied:
Yes(see [here](https://github.com/WordPress/gutenberg/pull/51477#discussion_r1235077297)). Could we invalidate the cache? I tried it a bit the first time, but couldn't..

youknowriad replied:
What is the "data" that is used in the resolver that was updated and when? Basically, we need to make sure the action that updates that piece of data returns "true" in `__experimentalGetTemplateForLink.shouldInvalidate`

ntsekouras replied:
That's great! I pushed a change for that. It's a shame that my first tries were to do just that, and because I didn't find a way to do it then, I circled around the proper solution for so long..

jameskoster commented:
> but still no way to have the preview(blocks) of that

It's interesting that comes up, because I've been (separately) looking at a more holistic update for the template panel that involves a "Reset" action for selecting the default template, rather than including it in the swap modal:

<img width="484" alt="Screenshot 2023-06-20 at 11 49 31" src="https://github.com/WordPress/gutenberg/assets/846565/a0154cc1-0dfe-40b4-beb2-8f7f0b89cda9">

I don't know if this makes things simpler or more complex on the code side, but perhaps something to consider?

ntsekouras commented:
@jameskoster I've updates the code to use the confirm swap template modal(option 2) and also used the design from your previous [comment](https://github.com/WordPress/gutenberg/pull/51477#issuecomment-1598544533).

jameskoster commented:
I pushed a change that moves the template detail to the Summary panel. This gets us a bit closer to what exists in the Post Editor and should help with familiarity. Two small issues remain:

1. The separator in the dropdown is the wrong color. I think this would be fixed by using `DropDownMenu` instead of `DropDown` but I couldn't get that working 😔
2. In the swap template flow, the confirmation modal appears _on top_ of the swap modal which feels a little busy. It may not work, but could we try replacing that modal instead, so that you only see one modal at once?

ntsekouras commented:
@jameskoster I pushed a commit to try the same modal. Since we are changing the inner content, the templates rendering needs to start over when we click `cancel` on the swap confirmation modal.

Regarding the DropDown styles, I copied some styles from the DropDownMenu component. I know it's not ideal but 🤷

jameskoster commented:
Thanks Nik, it's working well.

> I copied some styles from the DropDownMenu component. I know it's not ideal but 🤷

Would DropDownMenu not work here? I tried, but the props didn't seem to map over and my designer brain couldn't figure it out. Probably fine to address in a follow-up if it's not trivial.

There's one final change before I think we can merge. In the confirmation modal, the title is "Update [template name]", it should be "Update [page name]".

ntsekouras commented:
>Would DropDownMenu not work here?

I updated to use DropdownMenu and removed the style overrides. 

>There's one final change before I think we can merge. In the confirmation modal, the title is "Update [template name]", it should be "Update [page name]".

Now it's hardcoded as `Update page template?`. I don't get the suggestion.. Do you mean something like `Update template for: ${pageTitle}`?  If we insert the title in between other strings, it doesn't work well for i18n. Same goes if we would use the template name..

jameskoster commented:
I updated the title to "Save [page title]?". Does that work?

jameskoster reviewed (approved):
I think we're good in terms of design 🥳

youknowriad commented on packages/edit-site/src/components/sidebar-edit-mode/page-panels/reset-default-template.js:
```diff
@@ -0,0 +1,50 @@
+			<MenuItem
+				onClick={ async () => {
+					entitiy.edit( { template: '' }, { undoIgnore: true } );
+					await entitiy.save();
```
Is the saving necessary?

youknowriad replied:
The problem with this code is that it will save all the other entity changes that were pending, not just the template.

youknowriad reviewed (commented):
This is looking good to me overall. I just wonder if the template change should be saved directly. I saw some related discussions previously, what's the TLDR here?

ntsekouras commented:
>I just wonder if the template change should be saved directly. I saw some related discussions previously, what's the TLDR here?

We changed the message to say that the change will be published immediately, but we didn't account for the saving of other edits.. I mean the `undo` for them is still possible, but I agree it's not clear that this happens.

In order not to save I'd have to probably change `__experimentalGetTemplateForLink` not to make a request, but instead check the editedEntityRecord. I'm not sure if that could affect other flows/code. Any suggestions?

youknowriad commented:
In the post editor, you're able to edit the post template and click "edit" and it works without saving. It seems we may need a `core-data` function that says `getTemplateForPostType( postId )` or something like that and that function would take into consideration the edited "template" property otherwise fallback to `__experimentalGetTemplateForLink` or similar.

---

That said, maybe this PR is not the right place for it as it's a bit advanced so if you both think the current behavior is acceptable, I'm fine for this to be explored separately

jameskoster commented:
> but we didn't account for the saving of other edits

Apologies, it wasn't clear to me that _all_ unsaved changes are published – I assumed it was just the template change.

Given that realisation I don't think we can proceed with the current design. Instead of the "Save [page name]" confirmation modal we'd probably need to display the save modal from the site editor, this one:

<img width="628" alt="Screenshot 2023-06-22 at 11 00 11" src="https://github.com/WordPress/gutenberg/assets/846565/48508fd6-6989-451b-b668-4bb794f9ff82">

But I think that could be a bit unexpected in terms of the UX, so probably not worth rushing, especially we we've likely missed the 6.3 deadline anyway. 

I would still prefer that the template change **not** be published immediately, so perhaps it's worth looking into that now that we have more time?

jameskoster commented:
@ntsekouras I presume we'll need to punt this one?

ntsekouras commented:
>@ntsekouras I presume we'll need to punt this one?

Yes. It's not going to be in `6.3`. I'll pick it up again soon though.

ntsekouras committed:
[Site Editor - Page Inspector]: Add ability to switch templates

ntsekouras commented:
@jameskoster I've been trying to find a good solution for the swap template handling and so far I can't overcome the side effects from undo, if you swap one.

What I'm thinking is:
On `swap` button -> open the modal with the templates.
On clicking a template->udpate the site editor but **not** create an undo step and at the same time **not** save the changes.

Could that work? 

I haven't implemented the full solution for the above, but I think it's doable.

jameskoster commented:
Thanks for returning to this one.

So the issue is that we cannot add the template switch to the list of changes pending save _and_ have an undo item for swapping back?

I'd say adding the template switch to the list of unsaved changes is more important than the undo. It's relatively easy to 'undo' by simply switching back to the original template, and it's not a frequent action.

I think your suggestion sounds good, but it's a shame we can't have both. 

Edit: One additional thought which may or may not influence decision making here: after adding template swapping, we'll meed to add an option to create a _new_ template (potentially with options: `$custom` or `page-$slug`).

ntsekouras committed:
try setting the page/template from edited entity record

ntsekouras commented:
I pushed an update to use the edited `template` property, but there are so many use cases that I definitely need help in testing. Testing should include spotting any possible regression, theme switching etc..

For example if we have custom templates in pages and we switch theme, etc..

ntsekouras commented on packages/edit-site/src/store/actions.js:
```diff
@@ -257,22 +257,48 @@ export function setEditedPostContext( context ) {
 	async ( { dispatch, registry } ) => {
-		if ( ! page.path && page.context?.postId ) {
-			const entity = await registry
+		let template;
```
This is probably the most important changes that should be tested for any regressions or wrong handling in any use case.

jameskoster commented:
Given all the challenges we're up against here this feels like good progress.

I'm able to swap templates without triggering a save. However, when I do try to save I get a red "Saving failed." error notice. So unfortunately I can't test much further :D 

A couple of other small issues:

<img width="307" alt="Screenshot 2023-07-24 at 14 24 28" src="https://github.com/WordPress/gutenberg/assets/846565/aad68c5f-1cc4-4334-8a7f-691ba3b7bf44">

1. If the template name is long then the whole inspector gains a horizontal scrollbar.
2. The tooltip should probably say something like "Template options" rather than repeating the template name.

I'll see if I can tweak these two details.

jameskoster committed:
Update label

jameskoster committed:
Truncate

jameskoster commented:
Fixed the visual issues above:

<img width="279" alt="Screenshot 2023-07-24 at 14 49 35" src="https://github.com/WordPress/gutenberg/assets/846565/9977888f-cf6b-4bcd-919f-709fa02219bb">

ntsekouras commented:
>However, when I do try to save I get a red "Saving failed." error notice. So unfortunately I can't test much further :D

@jameskoster I don't have any errors 🤔 . Can you provide the error message or some reproduction steps? Also thank you for polishing!

jameskoster commented:
@ntsekouras it only seems to happen with the [Ollie](https://github.com/OllieWP/ollie) theme active 🧐

https://github.com/WordPress/gutenberg/assets/846565/a7761042-9d40-4b5a-b787-e4088d50756b

Other themes are fine. 

---

One issue I spotted with theme switching: If two themes have a template with the same name, the blocks of the previous theme will persist until you 'view' the template in the new theme:


https://github.com/WordPress/gutenberg/assets/846565/6df64e34-b20b-401d-a255-1ced9a4cceeb

annezazu commented:
@ntsekouras how are you feeling about this PR for 6.4? What do you need to keep this moving? Checking in as we are a month away from 6.4 beta 1 💥

jameskoster commented:
I'd love to see this added in 6.4 🙏

Given it's possible to create new pages and custom templates in the site editor, being unable to switch templates is a pretty big feature gap.

ntsekouras commented:
I've been working on different things lately, but I could give it some more iterations to see if it could land. In general this PR needs some code reviews for start, and also check the [specific case ](https://github.com/WordPress/gutenberg/pull/51477#issuecomment-1649940362)that doesn't work for @jameskoster - might be something wrong in the underlying code..

jameskoster commented:
Still seeing the issue as described, but only with the Ollie theme, so it may be theme related?

One other small issue is that the popover z-index seems to cause it to appear above the swap modal:


https://github.com/WordPress/gutenberg/assets/846565/24811212-95a3-47b6-9329-2084173f011d

ntsekouras committed:
update z-index of templates swap modal

ntsekouras committed:
fix typo

richtabor commented:
I'd like to see this land too. If we can get the technical aspect working, we can follow-up with tweaks to tighten up the experience.

ntsekouras commented:
I'll update/write tests in the morning, but it still needs code review 😄

ntsekouras committed:
add + update tests

ntsekouras commented:
Found the Ollie issue and will push the fix with a test. I wasn't constraining the results based on `postTypes` template property and that's why one of the fetched templates couldn't be saved - it had `postTypes: ['post']`.

ntsekouras committed:
fix suggestions per template's `postTypes` and test

scruffian commented:
This shouldn't stop the PR merging, but I'm not sure about the term "swap". It's not a term we use anywhere else, although it is a good one. We should probably align the term with the action we use to swap navigations in the navigation block for example...

scruffian commented on packages/edit-site/src/components/sidebar-edit-mode/page-panels/hooks.js:
```diff
@@ -0,0 +1,83 @@
+	const { postId } = useEditedPostContext();
+	return useSelect(
+		( select ) =>
+			+postId ===
```
what does the `+` do here?

ntsekouras replied:
Converts the `string` to `number`.

draganescu commented on packages/edit-site/src/components/sidebar-edit-mode/page-panels/style.scss:
```diff
@@ -36,15 +61,21 @@
 .edit-site-summary-field {
 	.components-dropdown {
-		flex-grow: 1;
+		width: 70%;
```
what is "70" :)

scruffian replied:
It's to stop this being full width:

<img width="315" alt="Screenshot 2023-09-12 at 11 06 52" src="https://github.com/WordPress/gutenberg/assets/275961/fcaa4078-f908-48b3-82c5-91d46f7468e7">

ntsekouras replied:
I'd defer to @jameskoster that made that styling change. I never asked 😄

draganescu commented on packages/edit-site/src/components/sidebar-edit-mode/page-panels/swap-template-button.js:
```diff
@@ -0,0 +1,82 @@
+		entitiy.edit( { template: template.name }, { undoIgnore: true } );
+		onClose();
+		onClick();
+		await setPage( {
```
I find it surprising that onClose and onClick ar note the last actions.

draganescu commented on packages/edit-site/src/store/actions.js:
```diff
@@ -257,22 +257,48 @@ export function setEditedPostContext( context ) {
-		if ( ! page.path && page.context?.postId ) {
-			const entity = await registry
+		let template;
+		const getDefaultTemplate = async ( slug ) =>
```
It's a bit mind bending that [here](https://github.com/WordPress/gutenberg/pull/51477/files#diff-345eaac452a15b557fde376f844090e4981ae5781716bf2acae186c87d63508aR25) the default template is an empty string and yet we still "get" it. Should it be a boolean like hasDefaultTemplate somewhere?

ntsekouras replied:
This is core WP here..

draganescu commented on packages/edit-site/src/store/actions.js:
```diff
@@ -257,22 +257,48 @@ export function setEditedPostContext( context ) {
+				} ),
+			} );
+
+		if ( page.path ) {
```
Why do we need this, can't we _always_ use `editedEntity?.template;`

ntsekouras replied:
This is when we don't have a selected page, where we show the home page.

MaggieCabrera commented:
> This shouldn't stop the PR merging, but I'm not sure about the term "swap". It's not a term we use anywhere else, although it is a good one. We should probably align the term with the action we use to swap navigations in the navigation block for example...

We use the term replace in other places like template parts:

<img width="618" alt="Screenshot 2023-09-12 at 12 02 32" src="https://github.com/WordPress/gutenberg/assets/3593343/e5071c3e-596f-41cb-a1ac-19ba08616e2d">

ntsekouras committed:
change order of calling in `onTemplateSelect`

SaxonF commented:
I may have missed the context around moving template into summary panel but I think it needs more prominence, partly because it's the primary place you switch to editing the template and we want that process to feel relatively low friction vs hiding in a dropdown menu. 


This also overlaps with https://href.li/?https://github.com/WordPress/gutenberg/pull/52674

scruffian reviewed (approved):
This is working well in my testing. I suggest we merge it before the PR gets any bigger and then create followup issues to make further improvements.

draganescu commented:
It is an apparently unrelated problem but if a page contains a ToC block swapping the template on this branch in the site edior kills Firefox.

ntsekouras commented:
>It is an apparently unrelated problem but if a page contains a ToC block swapping the template on this branch in the site edior kills Firefox.

That's the same with Chrome too.. We should pinpoint the problem though.. I'll cc @Mamaduka if has any ideas on top of his head..

youknowriad commented on packages/edit-site/src/components/sidebar-edit-mode/page-panels/hooks.js:
```diff
@@ -0,0 +1,83 @@
+ */
+import { store as editSiteStore } from '../../../store';
+
+export function useEditedPostContext() {
```
Nit: Do we really need this hook? Feels like a very light shortcut.

youknowriad replied:
I see that you're using this pattern for several hooks. Personally I wonder if this indirection is necessary to be honest.

youknowriad replied:
This pattern is going to increase the number of useSelect we have. While I didn't measure anything here. I wonder if a single useSelect is in general more performant than multiple ones. 

All of this is still a small thing though. I'm fine if you think this makes the code clearer/better.

ntsekouras replied:
I mostly did this because of the different dependencies of other `useSelect` in the code needed.

youknowriad commented on packages/edit-site/src/components/sidebar-edit-mode/page-panels/reset-default-template.js:
```diff
@@ -0,0 +1,44 @@
+		<MenuGroup>
+			<MenuItem
+				onClick={ async () => {
+					entity.edit( { template: '' }, { undoIgnore: true } );
```
Why are we using undoIngore in these edits?

ntsekouras replied:
Because the specific `undo` needs a side effect(to update the edited entity).

youknowriad replied:
I guess that means we need to solve that to solve the undo as well https://github.com/WordPress/gutenberg/pull/51477#discussion_r1323007695

youknowriad commented on packages/edit-site/src/components/sidebar-edit-mode/page-panels/reset-default-template.js:
```diff
@@ -0,0 +1,44 @@
+				onClick={ async () => {
+					entity.edit( { template: '' }, { undoIgnore: true } );
+					onClick();
+					await setPage( {
```
Is the idea here to cause a refresh of the template or something?

ntsekouras replied:
Exactly.

youknowriad commented on packages/edit-site/src/store/actions.js:
```diff
@@ -257,22 +257,48 @@ export function setEditedPostContext( context ) {
-
 		if ( ! template ) {
 			return;
 		}
```
I think the change in this action is fine, but what would be awesome (and probably better explored as a follow-up) is to remove the need for this action in favor of a `selector + resolver` that contains the same logic.

In other words when the editor component would calls these selectors:

```
getEditedEntityId()
getEditedEntityType()
```

a resolver is triggered to actually "resolve" the template (so the logic we have right now in this action).

The reason for this is that we would be able to say that `editEntityRecord` of the `template` actually marks the resolver as "unresolved" which would cause the resolver to trigger again. 

There are some subtle things to consider (loading...) that's why this is more of a quality of life improvement that should be addressed separately.

ntsekouras replied:
>The reason for this is that we would be able to say that editEntityRecord of the template actually marks the resolver as "unresolved" which would cause the resolver to trigger again.

Can you expand on this a bit?

youknowriad replied:
I mean that we can define the `shouldInvalidate` property on the resolver to ensure that every time the edit entity action is called with a new template, the corresponding resolver is invalidated.

youknowriad reviewed (commented):
I left some comments but nothing is blocking. The PR is very good code wise.

ntsekouras commented:
Regarding the TOC block issue, I think it's related to the specific block and doesn't work well if we have more headings in the template. I opened [an issue](https://github.com/WordPress/gutenberg/issues/54385) for that..

ntsekouras commented:
Let's land this and follow up with improvements. Thank you all!

jameskoster commented:
@SaxonF 

It then got moved to the Summary panel for a couple of reasons:

1. There was an issue where template swaps were saved instantly. This eliminated the idea of displaying a list of thumbnails in the Inspector as it felt very unnatural for those swaps to be immediately published.
2. A template panel featuring a single thumbnail felt a bit redundant – the template is already clearly visible on the canvas.
3. I'm not sure how common swapping a template is. Seems like something we might surface during initial page creation?
4. Similar for editing – there are direct pathways on the canvas.
5. Post editor consistency.

I don't think any of that prohibits a revisit though, especially if upcoming enhancements require it. I believe the instant-save issue get fixed along the way.

richtabor commented:
> I may have missed the context around moving template into summary panel but I think it needs more prominence, partly because it's the primary place you switch to editing the template and we want that process to feel relatively low friction vs hiding in a dropdown menu.

Agree, though I do like getting this in early for refinement. We could perhaps add a "Replace" button next to "Edit template", like how the draft/trash buttons are. A bit ad-hoc, but more appropriately positioned. 

<img width="351" alt="CleanShot 2023-09-12 at 17 19 22" src="https://github.com/WordPress/gutenberg/assets/1813435/b31a0a6e-a8a3-4c7f-9263-aa98d00db9bd">

SaxonF commented:
I don't want to block the work so happy to refine how we display the template in a follow up issue.

mikejolley commented:
This PR caused a regression with template lookup via the API. I created a patch here: https://github.com/WordPress/gutenberg/pull/54599

ntsekouras commented:
>This PR caused a regression with template lookup via the API. I created a patch here: https://github.com/WordPress/gutenberg/pull/54599

Thanks for catching this and fixing this @mikejolley! 🚀
